ext/intl/listformatter/listformatter_class.c

```diff
@@ -0,0 +1,241 @@
+
+    zend_object_std_init(&obj->zo, class_type);
+    object_properties_init(&obj->zo, class_type);
+    obj->zo.handlers = &listformatter_handlers;
+    return &obj->zo;
+}
+
+PHP_METHOD(IntlListFormatter, __construct)
+{
+    ListFormatter_object *obj = Z_INTL_LISTFORMATTER_P(ZEND_THIS);
+    char* locale;
+    size_t locale_len = 0;
+    zend_long type = ULISTFMT_TYPE_AND;
+    zend_long width = ULISTFMT_WIDTH_WIDE;
+    ZEND_PARSE_PARAMETERS_START(1, 3)
+        Z_PARAM_STRING(locale, locale_len)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_LONG(type)
+        Z_PARAM_LONG(width)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if(locale_len == 0) {
+        locale = (char *)intl_locale_get_default();
+    }
+
+    if (locale_len > INTL_MAX_LOCALE_LEN) {
+        zend_argument_value_error(1, "Locale string too long, should be no longer than %d characters", INTL_MAX_LOCALE_LEN);
+        RETURN_THROWS();
+    }
+
+    if (strlen(uloc_getISO3Language(locale)) == 0) {
+        zend_argument_value_error(1, "\"%s\" is invalid", locale);
+        RETURN_THROWS();
+    }
+
+    UErrorCode status = U_ZERO_ERROR;
+    #if U_ICU_VERSION_MAJOR_NUM >= 67
+        if (type != ULISTFMT_TYPE_AND && type != ULISTFMT_TYPE_OR && type != ULISTFMT_TYPE_UNITS) {
+            zend_argument_value_error(2, "must be one of IntlListFormatter::TYPE_AND, IntlListFormatter::TYPE_OR, or IntlListFormatter::TYPE_UNITS");
+            RETURN_THROWS();
+        }
+        
+        if (width != ULISTFMT_WIDTH_WIDE && width != ULISTFMT_WIDTH_SHORT && width != ULISTFMT_WIDTH_NARROW) {
+            zend_argument_value_error(3, "must be one of IntlListFormatter::WIDTH_WIDE, IntlListFormatter::WIDTH_SHORT, or IntlListFormatter::WIDTH_NARROW");
+            RETURN_THROWS();
+        }
+
+        LISTFORMATTER_OBJECT(obj) = ulistfmt_openForType(locale, type, width, &status);
+    #else
+        if (type != ULISTFMT_TYPE_AND) {
+            zend_argument_value_error(2, "contains an unsupported type. ICU 66 and below only support IntlListFormatter::TYPE_AND");
+            RETURN_THROWS();
+        }
+
+        if (width != ULISTFMT_WIDTH_WIDE) {
+            zend_argument_value_error(3, "contains an unsupported width. ICU 66 and below only support IntlListFormatter::WIDTH_WIDE");
+            RETURN_THROWS();
+        }
+
+        LISTFORMATTER_OBJECT(obj) = ulistfmt_open(locale, &status);
+    #endif
+
+    if (U_FAILURE(status)) {
+        intl_error_set(NULL, status, "Constructor failed", 0);
+        zend_throw_exception(IntlException_ce_ptr, "Constructor failed", 0);
+        RETURN_THROWS();
+    }
+}
+
+PHP_METHOD(IntlListFormatter, format)
+{
+    ListFormatter_object *obj = Z_INTL_LISTFORMATTER_P(ZEND_THIS);
+    zval *strings;
+
+    ZEND_PARSE_PARAMETERS_START(1, 1)
+        Z_PARAM_ARRAY(strings)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (!LISTFORMATTER_OBJECT(obj)) {
+        intl_error_set(NULL, U_ILLEGAL_ARGUMENT_ERROR, "ListFormatter not properly constructed", 0);
```

Comment:
We usually throw exceptions in case of uninitialized objects. i.e..:

```
zend_throw_exception(NULL,
			"The IntlIterator is not properly constructed", 0);
```

Reply:
Done